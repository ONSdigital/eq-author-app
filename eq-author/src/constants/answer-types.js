export const CHECKBOX = "Checkbox";
export const RADIO = "Radio";
export const TEXTFIELD = "TextField";
export const TEXTAREA = "TextArea";
export const CURRENCY = "Currency";
export const NUMBER = "Number";
export const PERCENTAGE = "Percentage";
export const DATE = "Date";
export const DATE_RANGE = "DateRange";
export const UNIT = "Unit";
export const DURATION = "Duration";

export const ROUTING_ANSWER_TYPES = [
  RADIO,
  CURRENCY,
  NUMBER,
  PERCENTAGE,
  CHECKBOX,
  UNIT,
];

export const RADIO_OPTION = "RadioOption";
export const CHECKBOX_OPTION = "CheckboxOption";
export const MUTUALLY_EXCLUSIVE = "MutuallyExclusive";
export const MUTUALLY_EXCLUSIVE_OPTION = "MutuallyExclusiveOption";

export default [
  CHECKBOX,
  RADIO,
  TEXTFIELD,
  TEXTAREA,
  CURRENCY,
  NUMBER,
  PERCENTAGE,
  DATE,
  DATE_RANGE,
  UNIT,
  DURATION,
];
