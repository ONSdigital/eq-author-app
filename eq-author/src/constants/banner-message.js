const bannerMessageMigrated =
  "Author has been migrated and this page will turn off shortly.";
const bannerMessageVersion = "Please use latest version of Author.";
const bannerMessageQuestions =
  "If you have any questions, please contact the Author team.";

const bannerMessages = [
  bannerMessageMigrated,
  bannerMessageVersion,
  bannerMessageQuestions,
];

export default bannerMessages;
