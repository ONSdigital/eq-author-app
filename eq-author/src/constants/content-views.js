export const ANSWER = "answer";
export const METADATA = "metadata";

export default [ANSWER, METADATA];
