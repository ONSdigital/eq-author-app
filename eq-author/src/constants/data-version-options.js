const DATA_VERSION_OPTIONS = [
  {
    value: "1",
    title: "Data version 1",
    description:
      "Enables downstream processing of older questionnaires without complex features",
  },
  {
    value: "3",
    title: "Data version 3",
    description:
      "Enables downstream processing of newer questionnaires with complex features",
  },
];

export default DATA_VERSION_OPTIONS;
