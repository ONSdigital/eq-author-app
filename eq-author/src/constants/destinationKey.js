export const EndOfQuestionnaire = "EndOfQuestionnaire";
export const NextPage = "NextPage";
export const Default = "Default";
export const EndOfCurrentSection = "EndOfCurrentSection";

export const destinationKey = {
  [EndOfQuestionnaire]: "End of questionnaire",
  [EndOfCurrentSection]: "End of current section",
  [NextPage]: "Next page",
  [Default]: "Select a destination",
};
