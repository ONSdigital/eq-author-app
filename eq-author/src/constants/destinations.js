export const EndOfQuestionnaire = "EndOfQuestionnaire";
export const NextPage = "NextPage";
export const Default = "Default";
export const EndOfCurrentSection = "EndOfCurrentSection";

export const logicalDestinations = (questionnaire) => [
  {
    id: NextPage,
    logicalDestination: NextPage,
    displayName: NextPage,
  },
  {
    id: EndOfCurrentSection,
    logicalDestination: EndOfCurrentSection,
    displayName: EndOfCurrentSection,
  },
  {
    id: EndOfQuestionnaire,
    logicalDestination: EndOfQuestionnaire,
    displayName: EndOfQuestionnaire,
    displayEnabled: !questionnaire.hub,
  },
];
