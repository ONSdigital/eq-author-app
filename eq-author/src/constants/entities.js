export const SECTION = "section";
export const PAGE = "page";
export const QUESTION_CONFIRMATION = "question-confirmation";
export const INTRODUCTION = "introduction";
export const METADATA = "metadata";
export const HISTORY = "history";
