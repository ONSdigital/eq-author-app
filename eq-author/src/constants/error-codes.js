export const ERR_PAGE_NOT_FOUND = "PAGE_NOT_FOUND";
