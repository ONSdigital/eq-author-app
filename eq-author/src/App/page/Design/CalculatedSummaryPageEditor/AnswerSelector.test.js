import React from "react";
import { shallow } from "enzyme";

import { NUMBER } from "constants/answer-types";

import {
  UnwrappedAnswerSelector as AnswerSelector,
  ErrorContext,
} from "./AnswerSelector";
import AnswerChip from "./AnswerChip";

describe("AnswerSelector", () => {
  let mockHandlers, page, answers;
  beforeEach(() => {
    mockHandlers = {
      fetchAnswers: jest.fn(),
      onChange: jest.fn(),
      onUpdate: jest.fn(),
      onChangeUpdate: jest.fn(),
      onUpdateCalculatedSummaryPage: jest.fn(),
      getValidationError: jest.fn(),
    };
    page = {
      id: "2",
      title: "",
      alias: "",
      pageType: "CalculatedSummaryPage",
      position: 2,
      displayName: "Foo",
      totalTitle: "",
      section: {
        id: "1",
        displayName: "This Section",
        questionnaire: { id: "1", metadata: [] },
      },
      summaryAnswers: [],
      availableSummaryAnswers: [],
    };
    answers = [
      {
        id: 1,
        displayName: "Answer 1",
        type: NUMBER,
        properties: {},
        page: {
          id: "1",
          displayName: "Foo",
          section: { id: 1, displayName: "Section 1" },
        },
      },
      {
        id: 2,
        displayName: "Answer 2",
        type: NUMBER,
        properties: {},
        page: {
          id: "1",
          displayName: "Foo",
          section: { id: 1, displayName: "Section 1" },
        },
      },
      {
        id: 3,
        displayName: "Answer 3",
        type: NUMBER,
        properties: {},
        page: {
          id: "1",
          displayName: "Foo",
          section: { id: 1, displayName: "Section 1" },
        },
      },
    ];
  });

  it("should render when there are no available summary answers", () => {
    const wrapper = shallow(<AnswerSelector page={page} {...mockHandlers} />);
    expect(wrapper).toMatchSnapshot();
  });

  it("should render when there are available summary answers", () => {
    page.availableSummaryAnswers = answers;
    const wrapper = shallow(<AnswerSelector page={page} {...mockHandlers} />);
    expect(wrapper).toMatchSnapshot();
  });

  it("should render a list of AnswerChips when answers selected", () => {
    page.summaryAnswers = answers;
    const wrapper = shallow(<AnswerSelector page={page} {...mockHandlers} />);
    expect(wrapper.find(AnswerChip)).toHaveLength(3);
  });

  it("should render an error when incompatible units selected", () => {
    mockHandlers.getValidationError = jest.fn(
      () => "Select answers that are the same unit type"
    );
    const wrapper = shallow(<AnswerSelector page={page} {...mockHandlers} />);
    expect(wrapper.find(ErrorContext)).toBeTruthy();
  });

  it("should remove single answers on remove click", () => {
    page.summaryAnswers = answers;
    const wrapper = shallow(<AnswerSelector page={page} {...mockHandlers} />);
    expect(wrapper.find(AnswerChip)).toHaveLength(3);
    wrapper.find(AnswerChip).first().simulate("remove");
    expect(mockHandlers.onUpdateCalculatedSummaryPage).toHaveBeenCalledWith({
      id: "2",
      summaryAnswers: [answers[1], answers[2]],
    });
  });

  it("should remove multiple answers on removeAll click", () => {
    page.summaryAnswers = answers;
    const wrapper = shallow(<AnswerSelector page={page} {...mockHandlers} />);
    expect(wrapper.find(AnswerChip)).toHaveLength(3);
    wrapper.find('[data-test="remove-all"]').simulate("click");
    expect(mockHandlers.onUpdateCalculatedSummaryPage).toHaveBeenCalledWith({
      id: "2",
      summaryAnswers: [],
    });
  });
});
