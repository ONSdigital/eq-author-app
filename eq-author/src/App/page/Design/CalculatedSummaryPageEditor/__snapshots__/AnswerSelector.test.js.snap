// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AnswerSelector should render when there are available summary answers 1`] = `
<AnswerSelector__Box>
  <AnswerSelector__Answers>
    <AnswerSelector__Empty>
      <AnswerSelector__EmptyTitle>
        No answers selected
      </AnswerSelector__EmptyTitle>
      <AnswerSelector__EmptyText>
        Select an answer using the button below.
      </AnswerSelector__EmptyText>
      <AnswerSelector__ErrorContainer>
        <AnswerSelector__EmptyButton
          data-test="answer-selector-empty"
          disabled={false}
          onClick={[Function]}
          small={true}
          type="button"
          variant="primary"
        >
          Select an answer
        </AnswerSelector__EmptyButton>
        <ErrorInline />
      </AnswerSelector__ErrorContainer>
    </AnswerSelector__Empty>
    <QuestionPicker
      data={
        Array [
          Object {
            "displayName": "Section 1",
            "folders": Array [
              Object {
                "alias": "Folder 1.1",
                "displayName": "Folder 1.1",
                "enabled": false,
                "id": "1.1",
                "pages": Array [
                  Object {
                    "additionalInfoContent": "",
                    "additionalInfoEnabled": false,
                    "additionalInfoLabel": "",
                    "alias": "1.1.1",
                    "answers": Array [
                      Object {
                        "displayName": "Answer 1",
                        "id": 1,
                        "properties": Object {},
                        "type": "Number",
                      },
                      Object {
                        "displayName": "Answer 2",
                        "id": 2,
                        "properties": Object {},
                        "type": "Number",
                      },
                      Object {
                        "displayName": "Answer 3",
                        "id": 3,
                        "properties": Object {},
                        "type": "Number",
                      },
                    ],
                    "definitionContent": "",
                    "definitionEnabled": false,
                    "definitionLabel": "",
                    "description": "",
                    "descriptionEnabled": false,
                    "displayName": "Page 1.1.1",
                    "folder": Object {
                      "id": "1",
                      "position": 0,
                    },
                    "guidance": "",
                    "guidanceEnabled": false,
                    "id": "1.1.1",
                    "pageType": "QuestionPage",
                    "position": 0,
                    "section": Object {
                      "displayName": "",
                      "id": "1",
                      "position": 0,
                    },
                    "title": "Page 1.1.1",
                    "validationErrorInfo": Object {
                      "errors": Array [],
                      "totalCount": 2,
                    },
                  },
                ],
                "position": 0,
                "section": Object {
                  "id": "1",
                },
                "validationErrorInfo": Object {
                  "__typename": "ValidationErrorInfo",
                  "errors": Array [],
                  "id": "1.1-validationErrorInfo",
                  "totalCount": 0,
                },
              },
            ],
            "id": "1",
            "position": 0,
            "title": "Section 1",
            "validationErrorInfo": Object {
              "totalCount": 0,
            },
          },
        ]
      }
      isOpen={false}
      onClose={[Function]}
      onSubmit={[Function]}
      showTypes={true}
      startingSelectedAnswers={Array []}
      title="Select one or more answer"
    />
  </AnswerSelector__Answers>
</AnswerSelector__Box>
`;

exports[`AnswerSelector should render when there are no available summary answers 1`] = `
<AnswerSelector__Box>
  <AnswerSelector__Answers>
    <AnswerSelector__Empty>
      <AnswerSelector__EmptyTitle>
        No answers available
      </AnswerSelector__EmptyTitle>
      <AnswerSelector__EmptyText>
        There are no answers to provide a calculated summary.
      </AnswerSelector__EmptyText>
      <AnswerSelector__ErrorContainer>
        <AnswerSelector__EmptyButton
          data-test="answer-selector-empty"
          disabled={true}
          onClick={[Function]}
          small={true}
          type="button"
          variant="primary"
        >
          Select an answer
        </AnswerSelector__EmptyButton>
        <ErrorInline />
      </AnswerSelector__ErrorContainer>
    </AnswerSelector__Empty>
    <QuestionPicker
      data={Array []}
      isOpen={false}
      onClose={[Function]}
      onSubmit={[Function]}
      showTypes={true}
      startingSelectedAnswers={Array []}
      title="Select one or more answer"
    />
  </AnswerSelector__Answers>
</AnswerSelector__Box>
`;
